File: java/src/main/java/problems1/RemoveNthNodeFromEndOfList.java
package problems1;

import concepts.SinglyListNode;

public enum RemoveNthNodeFromEndOfList {
    DEFAULT {
        @Override
        public SinglyListNode removeNthFromEnd(SinglyListNode head, int n) {
            if (head.next == null && n == 1) {
                return null;
            }
            SinglyListNode root = new SinglyListNode(0, head);
            SinglyListNode left = root;
            SinglyListNode right = root;
            for (int i = 0; i < n; i++) {
                left = left.next;
            }
            while (left.next != null) {
                left = left.next;
                right = right.next;
            }
            right.next = right.next.next;
            return root.next;
        }
    };

    public abstract SinglyListNode removeNthFromEnd(SinglyListNode head, int n);
}

File: kotlin/src/main/kotlin/problems1_100/RemoveNthNodeFromEndOfList.kt
package problems1_100

import concepts.SinglyListNode

enum class RemoveNthNodeFromEndOfList {
    DEFAULT {
        override fun removeNthFromEnd(head: SinglyListNode?, n: Int): SinglyListNode? {
            if (head?.next == null && n == 1) {
                return null
            }
            val root = SinglyListNode(0, head)
            var left: SinglyListNode? = root
            var right: SinglyListNode? = root
            for (i in 0 until n) {
                left = left!!.next
            }
            while (left!!.next != null) {
                left = left.next
                right = right!!.next
            }
            right!!.next = right.next!!.next
            return root.next
        }
    },
    ;

    abstract fun removeNthFromEnd(head: SinglyListNode?, n: Int): SinglyListNode?
}
